# 01: Importing libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Data Import

In [4]:
path = r'C:\Users\14239\Downloads\Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
# Create a dataframe

df_test = pd.DataFrame()

In [8]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

In [23]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The product_id column has a count of 49693, but a max of 49688. These numbers should be the same. We know that this is because of the 5 duplicate values that we already identified and removed, but we would want to investigate further when we first notice this.

I would also investigate the max price being 99999. That may be correct but the number seems alarmly high, and the mean is higher than the median, indicating that there potential for a mistake.

In [24]:
# Check for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed-type data identified by the function above.

In [25]:
# Check for missing values

df_ords.isnull().sum()

Unnamed: 0                           0
order_id                             0
user_id                              0
eval_set                             0
user_cumulative_order_number         0
orders_day_of_week                   0
order_hour_of_day                    0
days_since_prior_order          206209
dtype: int64

In [32]:
df_ords_clean = df_ords

In [33]:
# Adding new column first_order to signify a customer's first order

df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [34]:
df_ords_clean

Unnamed: 0  order_id  user_id eval_set  user_cumulative_order_number  \
0                 0   2539329        1    prior                             1   
1                 1   2398795        1    prior                             2   
2                 2    473747        1    prior                             3   
3                 3   2254736        1    prior                             4   
4                 4    431534        1    prior                             5   
...             ...       ...      ...      ...                           ...   
3421078     3421078   2266710   206209    prior                            10   
3421079     3421079   1854736   206209    prior                            11   
3421080     3421080    626363   206209    prior                            12   
3421081     3421081   2977660   206209    prior                            13   
3421082     3421082    272231   206209    train                            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         3                  7                    15.0   
2                         3                 12                    21.0   
3                         4                  7                    29.0   
4                         4                 15                    28.0   
...                     ...                ...                     ...   
3421078                   5                 18                    29.0   
3421079                   4                 10                    30.0   
3421080                   1                 12                    18.0   
3421081                   1                 12                     7.0   
3421082                   6                 14                    30.0   

         first_order  
0               True  
1              False  
2              False  
3              False  
4              False  
...              ...  
3421078        False  
3421079        False  
3421080        False  
3421081        False  
3421082        False  

[3421083 rows x 9 columns]

The missing data is still very important, since it signifies that that customer has no previous order, and therefore, is a first time customer. I added a new row to signify new customer.

In [35]:
# Check for duplicate values

df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [36]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, user_cumulative_order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

No duplicate rows found. 

In [37]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))